src/palace/manager/api/authentication/basic.py

```diff
@@ -716,39 +725,52 @@ def identifies_individuals(self):
         cls,
-        field: str | None,
+        value: str | None,
         restriction: str | list[str] | re.Pattern | None,
         match_type: LibraryIdentifierRestriction,
-    ) -> bool:
-        """Does the given patron match the given restriction?"""
-        if not field:
-            # No field -- nothing matches.
-            return False
+    ) -> tuple[bool, str]:
+        """Does the given patron match the given restriction?
 
-        if not restriction:
-            # No restriction -- anything matches.
-            return True
-
-        if match_type == LibraryIdentifierRestriction.REGEX:
-            if restriction.search(field):  # type: ignore[union-attr]
-                return True
-        elif match_type == LibraryIdentifierRestriction.PREFIX:
-            if field.startswith(restriction):  # type: ignore[arg-type]
-                return True
-        elif match_type == LibraryIdentifierRestriction.STRING:
-            if field == restriction:
-                return True
-        elif match_type == LibraryIdentifierRestriction.LIST:
-            if field in restriction:  # type: ignore[operator]
-                return True
-
-        return False
+        :param value: The value from the field we're matching against the restriction.
+        :param restriction: The restriction value.
+        :param match_type: The type of match we're performing.
+        :returns: True if the value matches the restriction, False otherwise.
+        """
+        value = value or ""
+
+        failure_reason = ""
+        match [match_type, restriction, value]:
+            case [LibraryIdentifierRestriction.NONE, *_]:
+                pass
+            case [_, _restriction, _] if not _restriction:
+                pass
+            case [_, _, _value] if _value is None or _value == "":
+                failure_reason = "No value in field."
+            case [LibraryIdentifierRestriction.REGEX, *_]:
+                if not (_pattern := cast(Pattern, restriction)).search(value):
+                    failure_reason = f"{value!r} does not match regular expression {_pattern.pattern!r}"
+            case [LibraryIdentifierRestriction.PREFIX, *_]:
+                if not value.startswith(_string := cast(str, restriction)):
+                    failure_reason = f"{value!r} does not start with {_string!r}"
+            case [LibraryIdentifierRestriction.STRING, *_]:
+                if value != (_string := cast(str, restriction)):
+                    failure_reason = f"{value!r} does not exactly match {_string!r}"
+            case [LibraryIdentifierRestriction.LIST, *_]:
+                if value not in (_list := cast(list, restriction)):
+                    failure_reason = f"{value!r} not in list {restriction!r}"
+
+        return (False, failure_reason) if failure_reason else (True, "")
 
     def get_library_identifier_field_data(
         self, patrondata: PatronData
     ) -> tuple[PatronData, str | None]:
```

Comment:
I think this change will need some corresponding changes to how the millennium patron auth works:
https://github.com/ThePalaceProject/circulation/blob/49c534bb6e012c06e60f16f4c0045a4243df40e6/src/palace/manager/api/millenium_patron.py#L154-L164

Reply:
Could you say a little more about this? I'll do some manual testing; but right now, I can't see the problem.

Reply:
I did some testing on this yesterday and it seemed to work as expected.